### PR TITLE
Refactor tool.cpp

### PR DESCRIPTION
**Pull request by wrrrzr**
Refactor tool.cpp using Guard Expressions and Extract Method

### CODE REVIEW

File: src/tool.cpp [hunk 1]
```diff
@@ -153,36 +167,46 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+		max_drop_level = root["max_drop_level"].asInt();
+
+	if (root["punch_attack_uses"].isInt())
+		punch_attack_uses = root["punch_attack_uses"].asInt();
+
+	deserializeJsonGroupcaps(root["groupcaps"]);
+	deserializeJsonDamageGroups(root["damage_groups"]);
+}
+
+void ToolCapabilities::deserializeJsonGroupcaps(Json::Value &object) {
```
Comment (JosiahWI) [inline src/tool.cpp:187]: This parameter is named `json` in other places, so let's keep it consistent. I think `json` is more descriptive anyway.

File: src/tool.cpp [hunk 2]
```diff
@@ -153,36 +167,46 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+	Json::ValueIterator iter;
+	ToolGroupCap value;
```
Comment (JosiahWI) [inline src/tool.cpp:189]: Could we declare these closer to their first uses?

File: src/tool.cpp [hunk 3]
```diff
@@ -153,36 +167,46 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+
+	if (!object.isObject())
+		return;
+
+	for (iter = object.begin(); iter != object.end(); ++iter) {
+		value.fromJson(*iter);
```
Comment (JosiahWI) [inline src/tool.cpp:195]: The name `value` doesn't seem very descriptive here. Maybe something like `groupCap`?

File: src/tool.cpp [hunk 4]
```diff
@@ -153,36 +167,46 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+		groupcaps[iter.key().asString()] = value;
+	}
+}
+
+void ToolCapabilities::deserializeJsonDamageGroups(Json::Value &object) {
```
Comment (JosiahWI) [inline src/tool.cpp:200]: Could we rename the parameter to `json`?

File: src/tool.cpp [hunk 5]
```diff
@@ -153,36 +167,46 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+	Json::ValueIterator iter;
+
+	if (!object.isObject())
+		return;
+
+	for (iter = object.begin(); iter != object.end(); ++iter) {
+		Json::Value &value = *iter;
```
Comment (JosiahWI) [inline src/tool.cpp:207]: Why is the pattern different from `deserializeJsonGroupcaps`?

Comment (JosiahWI) [reply src/tool.cpp:207]: Again, not a very descriptive name.

**Review by JosiahWI — CHANGES_REQUESTED** [review]
Thanks for applying these refactorings. They're pretty straightforward and this should have no behavioral change, but I'd like to clean up the extracted methods a little more.

**Comment by wrrrzr** [comment]
> Thanks for applying these refactorings. They're pretty straightforward and this should have no behavioral change, but I'd like to clean up the extracted methods a little more.

ok, I cleaned up the extracted methods. But I think it's better to call the variable in the loop "value" because it has a small scope.

File: src/tool.cpp (hunk 5 above)
Comment (JosiahWI) [reply src/tool.cpp:203]: It looks like `ToolGroupCap` is an object and `fromJson` is a method we defined on it, whereas this is just an integer.

**Review by JosiahWI — APPROVED** [review]
Looks good. Tested in MTG with bronze ax, cobble pick, and diamond pick: dig times for different nodes and tool uses still work.

File: src/tool.cpp [hunk 6]
```diff
@@ -153,36 +167,42 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+		max_drop_level = root["max_drop_level"].asInt();
+
+	if (root["punch_attack_uses"].isInt())
+		punch_attack_uses = root["punch_attack_uses"].asInt();
+
+	deserializeJsonGroupcaps(root["groupcaps"]);
+	deserializeJsonDamageGroups(root["damage_groups"]);
+}
+
+void ToolCapabilities::deserializeJsonGroupcaps(Json::Value &json) {
```
Comment (sfan5) [inline src/tool.cpp:187]:
```suggestion
void ToolCapabilities::deserializeJsonGroupcaps(Json::Value &json)
{
```

File: src/tool.cpp [hunk 7]
```diff
@@ -153,36 +167,42 @@ void ToolCapabilities::deserializeJson(std::istream &is)
...
+		return;
+
+	for (Json::ValueIterator iter = json.begin(); iter != json.end(); ++iter) {
+		ToolGroupCap value;
+		value.fromJson(*iter);
+		groupcaps[iter.key().asString()] = value;
+	}
+}
+
+void ToolCapabilities::deserializeJsonDamageGroups(Json::Value &json) {
```
Comment (sfan5) [inline src/tool.cpp:198]:
```suggestion
void ToolCapabilities::deserializeJsonDamageGroups(Json::Value &json)
{
```

File: src/tool.h [hunk 8]
```diff
@@ -83,6 +83,9 @@ struct ToolCapabilities
 	void deSerialize(std::istream &is);
 	void serializeJson(std::ostream &os) const;
 	void deserializeJson(std::istream &is);
+private:
```
Comment (sfan5) [inline src/tool.h:86]:
```suggestion

private:
```

Comment (wrrrzr) [reply src/tool.h:87]: > 

ok, i apply suggestions

**Review by sfan5 — APPROVED** [review]

**Review by appgurueu — APPROVED** [review]
Code is ok. I agree that the two variables named "value" could be named better but that isn't necessary IMO.

**Pull request merged**